esmero/strawberryfield
# Fixes deep nesting and improves performance

# FIXES

[BUG: json flattener is nesting too deep](https://github.com/esmero/strawberryfield/issues/16)

# HOW?

Some major improvements here, took me 2-3 hours (but writing this is again the biggest thing)
- Recursive common key flattening function was improved/fixed.
   * We don't over nest now and do some checking of everything that we accumulate.
   * We make sure we don't accumulate numeric keys (totally useless and also takes resources). Previously we were taking all first, second, third elements of everything and keeping them around...
- For each property we use to execute flattening again and again. Processing of that was moved to the field class itself and once that happens (once) every other call from an item internal can make use of the flattened array. This makes things in a large system way faster.
- Removed old class StrawberryDataByKeyProvider that was already replaced by StrawberryValuesFromJson (same functionality but better)
- If a computed property you decide to expose (imagine you request as:image) is a complex nested `Array` Solr won't know what to do with it. Now we jsonencode it (means it becomes a string). That way Solr can index it and if used together with a Solr Text field (try ngram!) you even get all values and keys tokenized and searchable. COOL!

# HOW TO TEST

Many options.
Go to this module's foler
```Shell
git fetch
git checkout fix-deepnesiting
```
and clear cache

1. test any node via the REST api (enable first, setup, see #15 for instructions). Example `http://localhost:8001/node/1?_format=json`. Generation should be quick and you should have all your values.
2. if you had trouble before and some fields did not made it into the Index, reindex your Solr, all your properties should now be indexed correctly (@giancarlobi `subject` will be OK)
```JSON
 "sm_subject":["digital library",
          "open access",
          "open-source",
          "repository"],
````

3. Now add to your strawberry_keynameprovider (in /admin/structure) a new key at the end of the coma separated values, use `subject_loc` (if you are using the webform provided you should have that one, if not well, you won't see anything, try any other key that has a deep nested elements). That field property (field descriptive metadata) should appear now in your Solr field config. Add it (if its not there, clear caches)as a field, set the field type to "Text" using the ngram option. Save. Reindex. Go to http://localhost:8983/solr/#/drupal/query and look for fields with `subject_loc` in its name like 
```JSON
 "tcngramm_X3b_en_subject_loc":["uri",
          "http",
          "idlocgov",
          "authorities",
          "subjects",
          "sh2001008470",
          "label",
          "Metrology"],

````

What you see there is a full hierarchy encoded as JSON and tokenized inside Solr. This allows even complex, deep nested elements to be "findable". Not really for faceting but yes, for discovery. This also means there is almost nothing (and well, there will be for sure once we use this more and more) that can break a Strawberryfield's internals (that deep white/pink fruity sweet flesh) exposure to Drupal and Solr.

@giancarlobi whenever you can give this a shot. I will wait for your feedback but will patch servers to be sure people see what they expect in the meantime.

@natehill once this is pull is merged (no need to test, i know its heavy on instructions)
please do a 
```Shell
docker exec -ti esmero-web bash -c 'composer update strawberryfield/strawberryfield' on your local deployment.
````

I hope this is the last time i have to deal with this particular issue. Seems we have now covered all the bases, we can for sure optimize and optimize to make things faster.
